api/query_engine.go

```diff
@@ -17,142 +19,326 @@ var (
+		// impact of big queries
+		// we could do two approaches: gradually reduce the interval of all series/groups being read, or just aggressively
+		// adjust one group at a time. The latter seems simpler, so for now we do just that.
+		if rp.PointsFetch() > uint32(mpprSoft) {
+			for group, split := range rp.pngroups {
+				if len(split.mdpno) > 0 {
+					split.mdpno, ok = planLowestResForMDPMulti(now, from, to, split.mdpno)
+					if !ok {
+						return nil, errUnSatisfiable
+					}
+					rp.pngroups[group] = split
+					if rp.PointsFetch() <= uint32(mpprSoft) {
+						goto HonoredSoft
+					}
+				}
+			}
+			for i, req := range rp.single.mdpno {
+				rp.single.mdpno[i], ok = planLowestResForMDPSingle(now, from, to, req)
+				if !ok {
+					return nil, errUnSatisfiable
+				}
+				// for every 10 requests we adjusted, check if we honor soft now.
+				// note that there may be thousands of requests
+				if i%10 == 9 {
+					if rp.PointsFetch() <= uint32(mpprSoft) {
+						goto HonoredSoft
+					}
+				}
+			}
+		}
+	}
+HonoredSoft:
+
+	if mpprHard > 0 && int(rp.PointsFetch()) > mpprHard {
+		return nil, errMaxPointsPerReq
 
-	for _, req := range reqs {
-		targets[req.Target] = struct{}{}
 	}
-	numTargets := uint32(len(targets))
-	minTTL := now - reqs[0].From
 
-	minIntervalSoft := uint32(0)
-	minIntervalHard := uint32(0)
+	// send out some metrics and we're done!
+	for _, r := range rp.single.mdpyes {
+		reqRenderChosenArchive.ValueUint32(uint32(r.Archive))
+	}
+	for _, r := range rp.single.mdpno {
+		reqRenderChosenArchive.ValueUint32(uint32(r.Archive))
+	}
+	for _, split := range rp.pngroups {
+		for _, r := range split.mdpyes {
+			reqRenderChosenArchive.ValueUint32(uint32(r.Archive))
+		}
+		for _, r := range split.mdpno {
+			reqRenderChosenArchive.ValueUint32(uint32(r.Archive))
+		}
+	}
+	reqRenderPointsFetched.ValueUint32(rp.PointsFetch())
+	reqRenderPointsReturned.ValueUint32(rp.PointsReturn(planMDP))
 
-	if maxPointsPerReqSoft > 0 {
-		minIntervalSoft = uint32(math.Ceil(float64(tsRange) / (float64(maxPointsPerReqSoft) / float64(numTargets))))
+	return &rp, nil
+}
+
+func planHighestResSingle(now, from, to uint32, req models.Req) (models.Req, bool) {
+	rets := getRetentions(req)
+	minTTL := now - from
+	var ok bool
+	for i, ret := range rets {
+		// skip non-ready option.
+		if ret.Ready > from {
+			continue
+		}
+		ok = true
+		req.Plan(i, ret)
+
+		if req.TTL >= minTTL {
+			break
+		}
 	}
-	if maxPointsPerReqHard > 0 {
-		minIntervalHard = uint32(math.Ceil(float64(tsRange) / (float64(maxPointsPerReqHard) / float64(numTargets))))
+	return req, ok
+}
+
+func planLowestResForMDPSingle(now, from, to uint32, req models.Req) (models.Req, bool) {
+	rets := getRetentions(req)
+	var ok bool
+	for i := len(rets) - 1; i >= 0; i-- {
+		// skip non-ready option.
+		if rets[i].Ready > from {
+			continue
+		}
+		ok = true
+		req.Plan(i, rets[i])
+		if req.PointsFetch() >= req.MaxPoints/2 {
```

Comment:
yes. perhaps we should compare against `max,req.maxPoints/2)` but then again we already know this an experimental feature and that this will be revised. So i'm not sure if i want to spend time reworking the MDP code as it will need a serious rethink.

Reply:
That's fair